sdk/resource/resource.go

```diff
@@ -51,29 +74,71 @@ func Merge(a, b *Resource) *Resource {
 	if b == nil {
 		return a
 	}
-	res := &Resource{
-		labels: map[core.Key]core.Value{},
+
+	n := len(a.sorted)
+	if len(b.sorted) > len(a.sorted) {
+		n = len(b.sorted)
+	}
+	// At a minimum the merge will be as large as the largest resource.
+	s := make([]core.KeyValue, 0, n)
+	k := make(map[core.Key]struct{}, n)
+	ai, bi := 0, 0
+	for ; ai < len(a.sorted) && bi < len(b.sorted); ai, bi = ai+1, bi+1 {
+		akv := a.sorted[ai]
+		s = append(s, akv)
+		k[akv.Key] = struct{}{}
+
+		bkv := b.sorted[bi]
+		if _, ok := a.keys[bkv.Key]; ok {
+			// a overwrites b.
+			continue
+		}
+		k[bkv.Key] = struct{}{}
+		s = append(s, bkv)
 	}
-	for k, v := range b.labels {
-		res.labels[k] = v
+
+	for ; ai < len(a.sorted); ai++ {
+		akv := a.sorted[ai]
+		s = append(s, akv)
+		k[akv.Key] = struct{}{}
 	}
-	// labels from resource a overwrite labels from resource b.
-	for k, v := range a.labels {
-		res.labels[k] = v
+
+	for ; bi < len(b.sorted); bi++ {
+		bkv := b.sorted[bi]
+		if _, ok := k[bkv.Key]; ok {
+			continue
+		}
+		k[bkv.Key] = struct{}{}
+		s = append(s, bkv)
 	}
-	return res
-}
 
-// Attributes returns a copy of attributes from the resource.
-func (r Resource) Attributes() []core.KeyValue {
-	attrs := make([]core.KeyValue, 0, len(r.labels))
-	for k, v := range r.labels {
-		attrs = append(attrs, core.KeyValue{Key: k, Value: v})
+	sort.Slice(s, func(i, j int) bool { return s[i].Key < s[j].Key })
+	res := &Resource{
+		keys:   k,
+		sorted: s,
 	}
-	return attrs
+	res.str = buildResourceString(res.sorted)
```

Comment:
This becomes unnecessary if string construction is deferred to invocation of `String()`.

```suggestion
```